### ISSUE TITLE
SIGSEGV when adding dropout layer

### ISSUE THREAD

**Issue by OleStauning** [tiny-dnn/tiny-dnn#540]
Code:
```
    network<sequential> net;
    adam opt;

    net << fc<relu>(9*2,40)
        << fc<relu>(40,10)
        << dropout(10,0.5)
        << fc<sigmoid>(10,2);

    int epoch=0;
    int epochs = 50;
    int batch = 50;
    net.fit<mse>(opt, xtrain, ytrain, batch, epochs,
    [](){},
    [&](){
        auto loss = net.get_loss<mse>(xtest, ytest);
        std::cout << "Loss=" << loss << std::endl;
    });
```

Error when using valgrind:
```
==7176== Invalid read of size 1
==7176==    at 0x115C35: tiny_dnn::dropout_layer::back_propagation(std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >*, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >*> > const&, std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >*, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >*> > const&, std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >*, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >*> >&, std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >*, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >*> >&) (dropout_layer.h:86)
==7176==    by 0x112683: tiny_dnn::layer::backward() (layer.h:515)
==7176==    by 0x11486E: tiny_dnn::sequential::backward(std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > > > const&) (nodes.h:285)
==7176==    by 0x13032C: void tiny_dnn::network<tiny_dnn::sequential>::bprop<tiny_dnn::mse>(std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > > > const&, std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > > > const&, std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > > > const&) (network.h:946)
==7176==    by 0x12D112: void tiny_dnn::network<tiny_dnn::sequential>::train_onebatch<tiny_dnn::mse, tiny_dnn::adam>(tiny_dnn::adam&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*, int, int, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*) (network.h:845)
==7176==    by 0x128B7C: void tiny_dnn::network<tiny_dnn::sequential>::train_once<tiny_dnn::mse, tiny_dnn::adam>(tiny_dnn::adam&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*, int, int, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*) (network.h:818)
==7176==    by 0x10E407: bool tiny_dnn::network<tiny_dnn::sequential>::fit<tiny_dnn::mse, tiny_dnn::adam, main::{lambda()#1}, main::{lambda()#2}>(tiny_dnn::adam&, std::vector<tiny_dnn::adam<tiny_dnn::adam<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<tiny_dnn::aligned_allocator> >, std::allocator<tiny_dnn::adam<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const&, std::vector<tiny_dnn::adam<tiny_dnn::adam<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<tiny_dnn::aligned_allocator> >, std::allocator<tiny_dnn::adam<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, unsigned long, int, main::{lambda()#1}, main::{lambda()#2}, bool, int, std::vector<tiny_dnn::adam<tiny_dnn::adam<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<tiny_dnn::aligned_allocator> >, std::allocator<tiny_dnn::adam<float, tiny_dnn::aligned_allocator<float, 64ul> > > >) (network.h:784)
==7176==    by 0x10E11B: bool tiny_dnn::network<tiny_dnn::sequential>::fit<tiny_dnn::mse, tiny_dnn::adam, main::{lambda()#1}, main::{lambda()#2}, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, tiny_dnn::aligned_allocator>(tiny_dnn::adam&, {lambda()#2}<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<tiny_dnn::adam> > const&, {lambda()#2}<tiny_dnn::aligned_allocator, std::allocator<{lambda()#2}<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<tiny_dnn::adam> > const> > const&, unsigned long, int, main::{lambda()#1}, main::{lambda()#2}, bool, int, {lambda()#2}<tiny_dnn::aligned_allocator, std::allocator<{lambda()#2}<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<tiny_dnn::adam> > const> >) (network.h:335)
.....................
==4475== Process terminating with default action of signal 11 (SIGSEGV)
==4475==  Access not within mapped region at address 0x10
==4475==    at 0x12113A: std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (shared_ptr_base.h:150)
==4475==    by 0x11AFCE: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() (shared_ptr_base.h:662)
==4475==    by 0x114D4F: std::__shared_ptr<tiny_dnn::edge, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() (shared_ptr_base.h:928)
==4475==    by 0x114D6B: std::shared_ptr<tiny_dnn::edge>::~shared_ptr() (shared_ptr.h:93)
==4475==    by 0x115E54: tiny_dnn::layer::backward() (layer.h:511)
==4475==    by 0x118078: tiny_dnn::sequential::backward(std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > > > const&) (nodes.h:285)
==4475==    by 0x136CC6: void tiny_dnn::network<tiny_dnn::sequential>::bprop<tiny_dnn::mse>(std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > > > const&, std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > > > const&, std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > > > const&) (network.h:946)
==4475==    by 0x13309E: void tiny_dnn::network<tiny_dnn::sequential>::train_onebatch<tiny_dnn::mse, tiny_dnn::adam>(tiny_dnn::adam&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*, int, int, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*) (network.h:845)
==4475==    by 0x12E01C: void tiny_dnn::network<tiny_dnn::sequential>::train_once<tiny_dnn::mse, tiny_dnn::adam>(tiny_dnn::adam&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*, int, int, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*) (network.h:818)
==4475==    by 0x110E71: bool tiny_dnn::network<tiny_dnn::sequential>::fit<tiny_dnn::mse, tiny_dnn::adam, main::{lambda()#1}, main::{lambda()#2}>(tiny_dnn::adam&, std::vector<tiny_dnn::adam<tiny_dnn::adam<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<tiny_dnn::aligned_allocator> >, std::allocator<tiny_dnn::adam<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const&, std::vector<tiny_dnn::adam<tiny_dnn::adam<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<tiny_dnn::aligned_allocator> >, std::allocator<tiny_dnn::adam<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, unsigned long, int, main::{lambda()#1}, main::{lambda()#2}, bool, int, std::vector<tiny_dnn::adam<tiny_dnn::adam<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<tiny_dnn::aligned_allocator> >, std::allocator<tiny_dnn::adam<float, tiny_dnn::aligned_allocator<float, 64ul> > > >) (network.h:784)
==4475==    by 0x110B85: bool tiny_dnn::network<tiny_dnn::sequential>::fit<tiny_dnn::mse, tiny_dnn::adam, main::{lambda()#1}, main::{lambda()#2}, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, tiny_dnn::aligned_allocator>(tiny_dnn::adam&, {lambda()#2}<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<tiny_dnn::adam> > const&, {lambda()#2}<tiny_dnn::aligned_allocator, std::allocator<{lambda()#2}<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<tiny_dnn::adam> > const> > const&, unsigned long, int, main::{lambda()#1}, main::{lambda()#2}, bool, int, {lambda()#2}<tiny_dnn::aligned_allocator, std::allocator<{lambda()#2}<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<tiny_dnn::adam> > const> >) (network.h:335)
```
The code works fine without the dropout layer. Am I doing something wrong?

Best regards

**Comment by smissan** [comment]
I saw a crash with dropout, but with a different model of course. Also works fine if I comment out dropout. This is on windows

**Comment by beru** [comment]
My guess is `out_type` of `dropout_layer` is `{vector_type::data}` and `in_type` of `fully_connected_layer` is `{vector_type::data, vector_type::weight, vector_type::bias}` or `{vector_type::data, vector_type::weight}`, so they can't be connected.

But if it's really so, `network` class should report an error about inappropriate layer configuration.

**Comment by beru** [comment]
It was my misunderstanding.

**Comment by beru** [comment]
In `dropout_layer::back_propagation` method, `in_grad` is updated.

```cpp
for (serial_size_t sample = 0;
      sample < static_cast<serial_size_t>(prev_delta.size()); ++sample) {
  for (serial_size_t i = 0;
        i < static_cast<serial_size_t>(curr_delta.size()); i++) {
    prev_delta[sample][i] = mask_[sample][i] * curr_delta[sample][i];
  }
}
```

During the process, `mask_` member is accesed. A problem is found by adding below assert statement.
```cpp
  assert(mask_[sample].size() == curr_delta.size());
```

**Comment by beru** [comment]
Posted a fix to the problem.
https://github.com/tiny-dnn/tiny-dnn/pull/543#issuecomment-282450220

**Comment by Randl** [comment]
@OleStauning can you please check with current master?

**Comment by OleStauning** [comment]
Yes - this have fixed the problem. Thanks.
